### PR TITLE
[fuzz] added fuzzer for network filter "ext_authz" 

### PR DESCRIPTION
**Pull request by jianwen612**
Commit Message:
Additional Description:
Added fuzz test code for network filter `ext_authz`. Added test input protobuf. Added two simple test cases.

Risk Level: Low
Testing:

After running for one minute with libfuzzer:
Line coverage for the .cc and .h files of `ext_authz` is 94.9 %
Function coverage for the .cc and .h files of `ext_authz` is 81.2% 

Running with libfuzzer(compiled with `opt` parameter), some output:

#[1001]   NEW    cov: 52728 ft: 68301 corp: 231/89Kb lim: 4096 exec/s: 250 rss: 263Mb L: 516/1547 MS: 1 CustomCrossOver-

#[2001]   NEW    cov: 53794 ft: 76654 corp: 358/194Kb lim: 4096 exec/s: 250 rss: 339Mb L: 2186/2280 MS: 5 CustomCrossOver-CustomCrossOver-CustomCrossOver-InsertRepeatedBytes-Custom-

#[3002]   NEW    cov: 54462 ft: 81081 corp: 447/283Kb lim: 4096 exec/s: 230 rss: 417Mb L: 1199/2486 MS: 7 Custom-ChangeBinInt-Custom-CopyPart-EraseBytes-ChangeByte-Custom-

#[5132]   NEW    cov: 54812 ft: 87536 corp: 575/441Kb lim: 4096 exec/s: 233 rss: 575Mb L: 1150/3625 MS: 4 Custom-CustomCrossOver-InsertRepeatedBytes-Custom-

/cc @asraa 
/cc @samkerner

### CODE REVIEW

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+          ON_CALL(*client, check(_, _, _, _))
+              .WillByDefault(WithArgs<0>(
+                  Invoke([&](Filters::Common::ExtAuthz::RequestCallbacks& callbacks) -> void {
+                    callbacks.onComplete(
+                        makeAuthzResponse(Filters::Common::ExtAuthz::CheckStatus::OK));
+                  })));
+          break;
+        }
+        case envoy::extensions::filters::network::ext_authz::Result::kCheckStatusError: {
+          ON_CALL(*client, check(_, _, _, _))
```
Comment (asraa) [inline test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc:76]: drive-by: could all of these three cases be combined with something like
`callbacks.onComplete(makeAuthzResponse(resultToCheckStatus(action.on_data().result())))`

Comment (jianwen612) [reply test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc]: > drive-by: could all of these three cases be combined with something like
> `callbacks.onComplete(makeAuthzResponse(resultToCheckStatus(action.on_data().result())))`

Good idea! Thanks. Updated.

File: test/extensions/filters/network/ext_authz/BUILD [hunk 2]
```diff
@@ -43,3 +45,27 @@ envoy_extension_cc_test(
...
+
+envoy_cc_fuzz_test(
+    name = "ext_authz_fuzz_test",
+    srcs = ["ext_authz_fuzz_test.cc"],
+    corpus = "ext_authz_corpus",
+    deps = [
+        ":ext_authz_fuzz_proto_cc_proto",
+        "//source/common/buffer:buffer_lib",
+        "//source/extensions/filters/network/ext_authz",
+        # "//test/fuzz:utility_lib",
```
Comment (asraa) [inline test/extensions/filters/network/ext_authz/BUILD:65]: nit: remove

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz.proto [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    option (validate.required) = true;
+    // Authorization check status
+    google.protobuf.Empty check_status_error = 1;
+    google.protobuf.Empty check_status_denied = 2;
+    google.protobuf.Empty check_status_ok = 3;
+  }
+}
+
+message OnData {
+  bytes data = 1 [(validate.rules).bytes = {min_len: 1}];
```
Comment (asraa) [inline test/extensions/filters/network/ext_authz/ext_authz_fuzz.proto:19]: is giving empty data a case that cannot occur? (does the filter manager check ahead of time?)

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  return check_status;
+}
+
+DEFINE_PROTO_FUZZER(const envoy::extensions::filters::network::ext_authz::ExtAuthzTestCase& input) {
+  try {
+    TestUtility::validate(input);
+  } catch (const ProtoValidationException& e) {
+    ENVOY_LOG_MISC(debug, "ProtoValidationException: {}", e.what());
+    return;
+  } catch (const ProtobufMessage::DeprecatedProtoFieldException& e) {
```
Comment (asraa) [inline test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc:63]: quick question -- why not catch any EnvoyException that throws from `TestUtility::validate`?

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  Filters::Common::ExtAuthz::MockClient* client = new Filters::Common::ExtAuthz::MockClient();
+  envoy::extensions::filters::network::ext_authz::v3::ExtAuthz proto_config = input.config();
+
+  ConfigSharedPtr config = std::make_shared<Config>(proto_config, stats_store);
+  std::unique_ptr<Filter> filter =
+      std::make_unique<Filter>(config, Filters::Common::ExtAuthz::ClientPtr{client});
+
+  NiceMock<Network::MockReadFilterCallbacks> filter_callbacks;
+  filter->initializeReadFilterCallbacks(filter_callbacks);
+  Network::Address::InstanceConstSharedPtr addr =
```
Comment (asraa) [inline test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc:78]: Can this be static? Will it improve exec/sec?

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    }
+    case envoy::extensions::filters::network::ext_authz::Action::kOnNewConnection: {
+      filter->onNewConnection();
+      break;
+    }
+    case envoy::extensions::filters::network::ext_authz::Action::kRemoteClose: {
+      filter_callbacks.connection_.raiseEvent(Network::ConnectionEvent::RemoteClose);
+      break;
+    }
+    case envoy::extensions::filters::network::ext_authz::Action::kLocalClose: {
```
Comment (asraa) [inline test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc:108]: Is there a contract about this? If the we have a LocalClose and then try to call onData, what happens?

File: test/extensions/filters/network/ext_authz/BUILD (hunk 2 above)
Comment (jianwen612) [reply test/extensions/filters/network/ext_authz/BUILD:65]: > nit: remove

Fixed. Thanks!

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz.proto (hunk 3 above)
Comment (jianwen612) [reply test/extensions/filters/network/ext_authz/ext_authz_fuzz.proto]: > is giving empty data a case that cannot occur? (does the filter manager check ahead of time?)

Empty data can occur. Fixed. Thanks!

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc (hunk 4 above)
Comment (jianwen612) [reply test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc:63]: > quick question -- why not catch any EnvoyException that throws from `TestUtility::validate`?

Fixed. Thanks!

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc (hunk 5 above)
Comment (jianwen612) [reply test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc]: > Can this be static? Will it improve exec/sec?

Set it to be static. I think it will improve to some extent, but no obvious improvement.

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc (hunk 6 above)
Comment (jianwen612) [reply test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc:110]: > Is there a contract about this? If the we have a LocalClose and then try to call onData, what happens?

LocalClose and RemoteClose will have the same effect on the filter: if there is a call which is being processed, the filter will cancel it. And stats info will change `config_->stats().active_.dec();` No other effect on following onData().

**Comment by asraa** [comment]
I think you may need to merge master to fix the coverage CI failure.

**Comment by jianwen612** [comment]
> I think you may need to merge master to fix the coverage CI failure.

Thanks! Merged.

**Comment by asraa** [comment]
@dio Can you take a quick look at this? It's a fuzzer we're writing for ext_authz network filter

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz.proto [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    // Connection close
+    google.protobuf.Empty remote_close = 3;
+    google.protobuf.Empty local_close = 4;
+  }
+}
+message ExtAuthzTestCase {
+  envoy.extensions.filters.network.ext_authz.v3.ExtAuthz config = 1
+      [(validate.rules).message = {required: true}];
+  repeated Action actions = 2;
+}
```
Comment (dio) [inline test/extensions/filters/network/ext_authz/ext_authz_fuzz.proto:41]: Super nit, add a newline.

File: test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+DEFINE_PROTO_FUZZER(const envoy::extensions::filters::network::ext_authz::ExtAuthzTestCase& input) {
+  try {
+    TestUtility::validate(input);
+  } catch (const ProtoValidationException& e) {
+    ENVOY_LOG_MISC(debug, "ProtoValidationException: {}", e.what());
+    return;
+  } catch (const ProtobufMessage::DeprecatedProtoFieldException& e) {
+    ENVOY_LOG_MISC(debug, "DeprecatedProtoFieldException: {}", e.what());
+    return;
+  } catch (const EnvoyException& e) {
```
Comment (asraa) [inline test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc:66]: @jianwen612 This might need some clean-up. There is no return statement here so the fuzzer would continue down the below block of code. 

However, we never end up failing on this line anyway since the exceptions thrown from validate are caught in the previous two catches. We could/should unify all these catches and catch EnvoyExceptions, I'm not sure if we need to have 3 separate behaviors for the catches since they'll all lead to early return (and we except to catch EnvoyExceptions)

Comment (jianwen612) [reply test/extensions/filters/network/ext_authz/ext_authz_fuzz_test.cc:66]: > @jianwen612 This might need some clean-up. There is no return statement here so the fuzzer would continue down the below block of code.
> 
> However, we never end up failing on this line anyway since the exceptions thrown from validate are caught in the previous two catches. We could/should unify all these catches and catch EnvoyExceptions, I'm not sure if we need to have 3 separate behaviors for the catches since they'll all lead to early return (and we except to catch EnvoyExceptions)

Thanks for pointing this out! I will fix this the next week!
